docs/proposals/env-vars.md

```diff
@@ -0,0 +1,316 @@
+type EnvVar struct {
+	// Name of the environment variable. Must be a C_IDENTIFIER.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Optional: no more than one of the following may be specified.
+
+	// Variable references $(VAR_NAME) are expanded
+	// using the previous defined environment variables in the container and
+	// any service environment variables. If a variable cannot be resolved,
+	// the reference in the input string will be unchanged. The $(VAR_NAME)
+	// syntax can be escaped with a double $$, ie: $$(VAR_NAME). Escaped
+	// references will never be expanded, regardless of whether the variable
+	// exists or not.
+	// Defaults to "".
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	// Source for the environment variable's value. Cannot be used if value is not empty.
+	// +optional
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.
+
+Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
+`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  
+
+With that, a new Shipwright type would be:
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
+   // after any Parameter variable substitution.
+   // +optional
+   DefaultValue  string
+}
+```
+
+Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 
+updated with an array of `EnvironmentVariable`
+
+`BuildStrategy` would also have a `EnvironmentVariable` array named `DisallowedEnvironmentVariables` that lists `EnvVar` settings which are not allowed
+and will be filtered in the final application onto the `TaskSpec`.
+
+With settings at each of those "layers", order of precedence comes in.
+
+A proposed default would be "developer centric":
+- apply `BuildStrategy` envs first
+- then `Build`
+- then `BuildRun`
+
+Then, via the [current global config of controller env vars](https://github.com/shipwright-io/build/blob/v0.4.0/docs/configuration.md),
+or if that current approach gets converted to a `ConfigMap`, or if the community decides to create a `ConfigMap` for each
+logical piece of "feature configuration", we have a setting that allows a switch to "ADMINISTRATOR_CENTRIC" order of precedence,
+say:
+
+- apply `BuildRun` envs first
+- then `Build`
+- then `BuildStrategy`
+
+
```

Comment:
Looking for some more information on where we set those environment variables eventually. I assume on all steps defined in the build strategy = not including those steps/containers that we dynamically define like to retrieve sources ?

Reply:
Yeah I can add more details here based on how we resolve items above after our conversation.